lib/provider/src/provider.ts

```diff
@@ -1,31 +1,39 @@
 import type {
     AnnotationsParams,
     AnnotationsResult,
+    CandidateItemsParams,
+    CandidateItemsResult,
     CapabilitiesParams,
     CapabilitiesResult,
     ItemsParams,
     ItemsResult,
     ProviderSettings,
 } from '@openctx/protocol'
+import type { ItemKindID } from '@openctx/schema'
 
 /**
  * An OpenCtx provider implemented in TypeScript/JavaScript.
  *
  * @template S The type of provider settings.
  */
-export interface Provider<S extends {} = ProviderSettings> {
+export interface Provider<S extends {} = ProviderSettings, K extends ItemKindID = ItemKindID> {
     /**
      * Reports the capabilities of the provider.
      */
     capabilities(
         params: CapabilitiesParams,
         settings: S
-    ): CapabilitiesResult | Promise<CapabilitiesResult>
+    ): CapabilitiesResult<K> | Promise<CapabilitiesResult<K>>
+
+    candidateItems?(
+        params: CandidateItemsParams<K>,
+        settings: S
+    ): CandidateItemsResult | Promise<CandidateItemsResult>
 
     /**
      * Returns OpenCtx items.
      */
-    items?(params: ItemsParams, settings: S): ItemsResult | Promise<ItemsResult>
+    items?(params: ItemsParams<K>, settings: S): ItemsResult | Promise<ItemsResult>
```

Comment:
it's a little unclear to me the value this generic adds. Can you motivate the benefits for my noobness.

Reply:
While the generic will go away as the schema.ts will be generated from json and will provide no generics I think. 

The benefit of having a generic is that, while writing a provider in TS, we get a bit of type safety, as follows:

```typescript
type GithubItemKind = 'pr' | 'issue'
const MyProvider: <{}, 'pr' | 'issue'> = {
  items({kind}) {
    if (kind == 'pr') {
      return {
        kind: 'pr', // this will be typesafe and only allow either 'pr' or 'issue' strings. 
      }
    }
  }
}
```

But anyways, it will be removed.

Reply:
Gone are the generics.